[export] Support module inputs for non strict mode.

zhxchen17 opened the pull request:
Summary:
Add experimental support for torch.nn.Module as input types.

Before this change, we don't support module inputs but recently we saw some interesting use cases like gpt-fast https://github.com/pytorch-labs/gpt-fast/blob/main/generate.py#L68 where we directly pass in a module input for different variants of the same models.

Since we don't really care about non-param or non-buffer states in non strict mode, we don't care about those either and pretend they are like plain constants during tracing. We treat any module input like a nested container of tensor, and each time we will automatically register a pytree handler for these module types to flatten its state dict into a group of tensors. We will just inline any module method call during tracing like we did for `self` module in export_for_training. This will make input modules' behavior very similar to the training module in typical case, except that we don't record the inputs as parameter or buffers but rather just plain user inputs.

Test Plan: buck run mode/opt caffe2/test:test_export -- -r test_module_input

Differential Revision: D67680827

tugsbayasgalan commented on torch/_export/utils.py:
```diff
@@ -1150,3 +1153,146 @@ def _compiling_state_context():
+        assert flatten_fn is not None
+        flattened, _ = flatten_fn(obj)
+
+        def copy_module(mod: torch.nn.Module):
```
Can you add a detailed note explaining what you are doing here? What i am seeing here is you are probably creating a dummy nn.Module using the old state dict values?

tugsbayasgalan commented on torch/_export/utils.py:
```diff
@@ -1150,3 +1153,146 @@ def _compiling_state_context():
+
+def register_module_as_pytree_input_node(
+    cls: Type[torch.nn.Module],
+    flatten_fn: Optional[FlattenFunc] = None,
```
Looks like all these optional params are actually not used. Could we make this method simpler?

tugsbayasgalan commented on torch/_export/non_strict_utils.py:
```diff
@@ -446,6 +446,42 @@ def inner(m: torch.nn.Module, prefix_atoms: List[str], constants):
+    ctxs = []
+    for arg in pytree.tree_leaves((args, kwargs)):
+        if isinstance(arg, torch.nn.Module):
+            from torch.export._trace import _fakify_params_buffers
```
Hmm should we move this to a util file? `from torch.export._trace import _fakify_params_buffers`

bdhirsh commented on torch/_export/passes/lift_constants_pass.py:
```diff
@@ -199,7 +199,7 @@ def lift_constants_pass(
                 if isinstance(constant_val, torch.nn.Parameter):
                     warnings.warn(
-                        f"{node.target} created when tracing {node.meta['stack_trace']} is a parameter. But"
+                        f"{node.target} created when tracing {node.meta.get('stack_trace', '<unknown stack>')} is a parameter. But"
```
nit: this seems better than crashing for sure. But if you also found an example test where we found a param with no stacktrace, should we also try to fix that case (so we get a proper stacktrace?)

zhxchen17 replied:
this error was found in my previous version of the diff, and couldn't be reproduced with the same example I have with the current version... I think the reason is that in the buggy version, we produced some extra node in the graph which doesn't have the stacktrace. Nonetheless, I guess I will leave the code here as it's better to not crash with this warning.

bdhirsh commented on torch/_export/utils.py:
```diff
@@ -1150,3 +1153,146 @@ def _compiling_state_context():
+            return s
+
+    _register_pytree_node(
+        cls,
```
will this permanently make the user's specific nn module subclass pytree-able for the rest of the live process? If it does, you probably want to "de-register" at the end of export so there are no user-visible side effects.

E.g.:
```
# before, this should not flatten the user's mod
mod_and_flattened_tensors, _ = pytree.tree_flatten(mod, tensor_list)

_ = export.export(mod, ...)

# After exporting, the behavior of pytree on the user code should be unchanged
mod_and_flattened_tensors, _ = pytree.tree_flatten(mod, tensor_list)


```

zhxchen17 replied:
@bdhirsh yeah in principle I guess agree with you, while for testing purposes, I need to make sure ep.module() also works after torch.export, so I either to
1. make the side effect kept there
2. re-register and re-de-resgister the module types in .module() again
3. hide this behavior behind a flag/context manager so that user has to opt in this behavior.

Not sure which one sounds better, any opinion?

bdhirsh replied:
hmm I think (2) overall sounds like a better outcome than the global side effect. Do you think it's something we can implement in a reasonable / clean way?

zhxchen17 replied:
yeah I can give it a try.

bdhirsh commented on torch/export/_trace.py:
```diff
@@ -1411,6 +1413,7 @@ def _strict_export_lower_to_aten_ir(
 
     return ExportArtifact(
         aten=aten_export_artifact,
+        in_spec=orig_in_spec,
```
Does `ExportedProgram.module()` work properly when you have nn module graph inputs? (I guess it now needs to know to unflatten some tensor inputs back into an nn module)

zhxchen17 commented on test/export/test_export.py:
```diff
@@ -5535,6 +5563,32 @@ def forward(self, x):
+        ref_x = torch.randn(2, 2)
+        ref_out = f(ref_x, mod)
+
+        ep = torch.export.export_for_training(f, (torch.randn(2, 2), mod), strict=False)
```
@tugsbayasgalan I don't think strict mode is prioritized here.

avikchaudhuri commented:
I'm not entirely sure what the invariants are here. Exporting with a module input of class `M` seems to mean, different state dicts for `M()` can be passed in but the same code `M.forward()` would be run. Is that correct? So if later a module input of class `N` is passed, (a) the state dict has to have the same structure (b) any code difference for the forward will be ignored.

Can you confirm? If so, the behavior feels pretty weird / surprising to me (although if you are asserting the input module class is specialized and must be the same between export and later, then it feels less weird). Otherwise please clarify what I'm getting wrong...

bdhirsh commented on torch/_export/non_strict_utils.py:
```diff
@@ -446,6 +447,54 @@ def inner(m: torch.nn.Module, prefix_atoms: List[str], constants):
 
 
+@contextlib.contextmanager
+def _enable_module_inputs(
```
maybe a dumb nit: I parsed this name as "inputs to the exported program module", which is wrong, since it's really "inputs to the exported program that are themselves nn.modules"

name bikeshedding.... `_enable_graph_inputs_of_type_nn_module`?

zhxchen17 committed:
[export] Support module inputs for non strict mode. (#143925)

Summary:

Add experimental support for torch.nn.Module as input types.

Before this change, we don't support module inputs but recently we saw some interesting use cases like gpt-fast https://github.com/pytorch-labs/gpt-fast/blob/main/generate.py#L68 where we directly pass in a module input for different variants of the same models.

Since we don't really care about non-param or non-buffer states in non strict mode, we don't care about those either and pretend they are like plain constants during tracing. We treat any module input like a nested container of tensor, and each time we will automatically register a pytree handler for these module types to flatten its state dict into a group of tensors. We will just inline any module method call during tracing like we did for `self` module in export_for_training. This will make input modules' behavior very similar to the training module in typical case, except that we don't record the inputs as parameter or buffers but rather just plain user inputs.

Test Plan: buck run mode/opt caffe2/test:test_export -- -r test_module_input

Differential Revision: D67680827

tugsbayasgalan commented:
> > I'm not entirely sure what the invariants are here. Exporting with a module input of class `M` seems to mean, different state dicts for `M()` can be passed in but the same code `M.forward()` would be run. Is that correct? So if later a module input of class `N` is passed, (a) the state dict has to have the same structure (b) any code difference for the forward will be ignored.
> > Can you confirm? If so, the behavior feels pretty weird / surprising to me (although if you are asserting the input module class is specialized and must be the same between export and later, then it feels less weird). Otherwise please clarify what I'm getting wrong...
> 
> @avikchaudhuri We cannot pass in a module input with class `N` because it violates type safety and it should be already caught by pytree flattening (which will detect whether N and M are the same type during flattening). So yeah, we are asserting input module must be the same type here.
> 
> I think in typical case (like gpt-fast) this seems pretty useful when you are doing inference with different variants of the transformer models, like quantization, or even speculative decoding. Therefore I think the motivation here is pretty strong.

To follow up on this, does this mean we also error on subclasses of model `N` ?

zhxchen17 commented:
> > > I'm not entirely sure what the invariants are here. Exporting with a module input of class `M` seems to mean, different state dicts for `M()` can be passed in but the same code `M.forward()` would be run. Is that correct? So if later a module input of class `N` is passed, (a) the state dict has to have the same structure (b) any code difference for the forward will be ignored.
> > > Can you confirm? If so, the behavior feels pretty weird / surprising to me (although if you are asserting the input module class is specialized and must be the same between export and later, then it feels less weird). Otherwise please clarify what I'm getting wrong...
> > 
> > 
> > @avikchaudhuri We cannot pass in a module input with class `N` because it violates type safety and it should be already caught by pytree flattening (which will detect whether N and M are the same type during flattening). So yeah, we are asserting input module must be the same type here.
> > I think in typical case (like gpt-fast) this seems pretty useful when you are doing inference with different variants of the transformer models, like quantization, or even speculative decoding. Therefore I think the motivation here is pretty strong.
> 
> To follow up on this, does this mean we also error on subclasses of model `N` ?

@tugsbayasgalan yes, module subclasses are not supported because most of the time it doesn't make sense. (instead we should encourage users to use tensor subclasses)
